worker/upload.go

```diff
@@ -36,173 +31,91 @@ const (
-		up.rs = rs
-	}
-}
-
 type (
-	slabID [8]byte
-
 	uploadManager struct {
-		b      Bus
-		hp     hostProvider
-		rl     revisionLocker
-		logger *zap.SugaredLogger
-		mm     *memoryManager
+		b           Bus
+		hp          hostProvider
+		rl          revisionLocker
+		mm          memoryManager
+		logger      *zap.SugaredLogger
+		shutdownCtx context.Context
 
 		maxOverdrive     uint64
 		overdriveTimeout time.Duration
 
 		statsOverdrivePct              *stats.DataPoints
 		statsSlabUploadSpeedBytesPerMS *stats.DataPoints
-		stopChan                       chan struct{}
 
-		mu            sync.Mutex
-		uploaders     []*uploader
-		lastRecompute time.Time
+		mu        sync.Mutex
+		uploaders []*uploader
 	}
 
 	uploader struct {
-		mgr *uploadManager
+		b Bus
 
-		hk         types.PublicKey
-		siamuxAddr string
+		hk              types.PublicKey
+		siamuxAddr      string
+		signalNewUpload chan struct{}
+		shutdownCtx     context.Context
 
-		statsSectorUploadEstimateInMS    *stats.DataPoints
-		statsSectorUploadSpeedBytesPerMS *stats.DataPoints // keep track of this separately for stats (no decay is applied)
-		signalNewUpload                  chan struct{}
-		stopChan                         chan struct{}
-
-		mu                  sync.Mutex
-		host                hostV3
-		fcid                types.FileContractID
-		renewedFrom         types.FileContractID
-		endHeight           uint64
-		bh                  uint64
+		mu        sync.Mutex
+		bh        uint64
+		endHeight uint64
+		fcid      types.FileContractID
+		host      hostV3
+		queue     []*sectorUploadReq
+
+		// stats related field
 		consecutiveFailures uint64
-		queue               []*sectorUploadReq
+		lastRecompute       time.Time
+
+		statsSectorUploadEstimateInMS    *stats.DataPoints
+		statsSectorUploadSpeedBytesPerMS *stats.DataPoints
 	}
 
-	upload struct {
-		id  api.UploadID
-		mgr *uploadManager
+	uploadManagerStats struct {
+		avgSlabUploadSpeedMBPS float64
+		avgOverdrivePct        float64
+		healthyUploaders       uint64
+		numUploaders           uint64
+		uploadSpeedsMBPS       map[types.PublicKey]float64
+	}
 
-		allowed      map[types.FileContractID]struct{}
+	upload struct {
+		id           api.UploadID
+		allowed      map[types.PublicKey]struct{}
 		lockPriority int
-
-		mu   sync.Mutex
-		used map[slabID]map[types.FileContractID]struct{}
+		shutdownCtx  context.Context
 	}
 
 	slabUpload struct {
-		mgr    *uploadManager
-		mem    *acquiredMemory
-		upload *upload
+		uploadID         api.UploadID
+		created          time.Time
+		lockPriority     int
+		maxOverdrive     uint64
+		mem              *acquiredMemory
+		overdriveTimeout time.Duration
 
-		sID     slabID
-		created time.Time
-		shards  [][]byte
+		candidates []*uploader // sorted by upload estimate
+		shards     [][]byte
 
 		mu          sync.Mutex
 		numInflight uint64
 		numLaunched uint64
+		numUploaded uint64
+
+		overdriving    map[int]map[types.PublicKey]struct{}
```

Comment:
Hm I had it at one point but decided to move it because it gets rid of the only lock necessary on `sectorUpload`. I hate the double map here too so I'll move it back. We need to guard it with a mutex though because of overdrive. I'm wondering though if we can change the loop so we don't have to overdrive in a separate thread... that'd be better I think I'll try that